packages/odyssey-react/src/components/TextInput/TextInput.test.tsx

```diff
@@ -0,0 +1,101 @@
+/*!
+ * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import React from "react";
+import { render, fireEvent, within } from "@testing-library/react";
+import TextInput from ".";
+import type { TextInputProps } from ".";
+
+const textBox = 'textbox';
+const label = 'Destination';
+
+describe("TextInput", () => {
+  it('renders into the document', () => {
+    const { getByRole } = render(
+      <TextInput label={label} />
+    );
+
+    expect(getByRole(textBox)).toBeInTheDocument();
+  });
+
+  it('renders a provided id associating the input and label', () => {
+    const { getByRole, getByText } = render(
+      <TextInput label={label} id="foo" />
+    );
+
+    expect(getByRole(textBox)).toHaveAttribute('id', 'foo');
+    expect(getByText(label)).toHaveAttribute('for', 'foo');
+  });
+
+  it.each([
+    ['disabled'],
+    ['readonly'],
+    ['required']
+  ])('renders %p attribute', (attr: string) => {
+    const { getByRole } = render(
+      <TextInput label={label} {...{ [attr]: true }} />
+    );
+
+    expect(getByRole(textBox)).toHaveAttribute(attr);
+  });
+
+  it.each<[TextInputProps['type']]>([
+    [undefined],
+    ['text'],
+    ['email'],
+    ['url'],
+    ['tel'],
+    ['search'],
+    ['password']
+  ])('renders %p input type', (type) => {
+    const { getByLabelText } = render(
+      <TextInput label={label} type={type} />
+    );
+
+    expect(getByLabelText(label)).toHaveAttribute('type', type ?? 'text');
+  });
+
+  it('renders the optionalLabel when input is not required', () => {
+    const optionalLabel = 'Optional';
+
+    const { container } = render(
+      <TextInput label={label} optionalLabel={optionalLabel} required={false} />
```

Comment:
On the UX side, it's best practice to treat fields as required by default, since that's the most common use case. It's why we indicate fields as optional rather than indicating them as required.